rclcpp/include/rclcpp/mapped_ring_buffer.hpp

```diff
@@ -101,53 +102,91 @@ class MappedRingBuffer : public MappedRingBufferBase
    * \param value if the key is found, the value is stored in this parameter
    */
   void
-  get_copy_at_key(uint64_t key, ElemUniquePtr & value)
+  get(uint64_t key, ElemUniquePtr & value)
   {
     std::lock_guard<std::mutex> lock(data_mutex_);
     auto it = get_iterator_of_key(key);
     value = nullptr;
     if (it != elements_.end() && it->in_use) {
-      auto ptr = ElemAllocTraits::allocate(*allocator_.get(), 1);
-      ElemAllocTraits::construct(*allocator_.get(), ptr, *it->value);
-      value = ElemUniquePtr(ptr);
+      T * raw_ptr;
+      ElemDeleter * deleter;
+      if (it->unique_value) {
+        raw_ptr = it->unique_value.get();
+        deleter = &it->unique_value.get_deleter();
+      } else {
+        raw_ptr = const_cast<T *>(it->shared_value.get());
+        deleter = std::get_deleter<ElemDeleter, const T>(it->shared_value);
+      }
+      if (raw_ptr) {
+        auto ptr = ElemAllocTraits::allocate(*allocator_.get(), 1);
+        ElemAllocTraits::construct(*allocator_.get(), ptr, *raw_ptr);
+        if (deleter) {
+          value = ElemUniquePtr(ptr, *deleter);
+        } else {
+          value = ElemUniquePtr(ptr);
+        }
+      }
     }
   }
 
-  /// Return ownership of the value stored in the ring buffer, leaving a copy.
+  /// Return a shared_ptr of the value stored in the ring buffer at the given key.
   /**
-   * The key is matched if an element in the ring bufer has a matching key.
-   * This method will allocate in order to store a copy.
+   * The key is matched if an element in the ring buffer has a matching key.
+   * This method will allocate in order to return a copy.
    *
    * The key is not guaranteed to be unique, see the class docs for more.
    *
-   * The ownership of the currently stored object is returned, but a copy is
-   * made and stored in its place.
-   * This means that multiple calls to this function for a particular element
-   * will result in returning the copied and stored object not the original.
-   * This also means that later calls to pop_at_key will not return the
-   * originally stored object, since it was returned by the first call to this
-   * method.
+   * The contents of value before the method is called are discarded.
+   *
+   * \param key the key associated with the stored value
+   * \param value if the key is found, the value is stored in this parameter
+   */
+  void
+  get(uint64_t key, ElemSharedPtr & value)
+  {
+    std::lock_guard<std::mutex> lock(data_mutex_);
+    auto it = get_iterator_of_key(key);
+    value.reset();
+    if (it != elements_.end() && it->in_use) {
+      if (!it->shared_value) {
+        it->shared_value = std::move(it->unique_value);
+      }
+      value = it->shared_value;
+    }
+  }
+
+  /// Return a copy of the stored element and reset the stored shared_ptr.
```

Comment:
This appears to be returning ownership, rather than a copy.  Please update the documentation to reflect this.